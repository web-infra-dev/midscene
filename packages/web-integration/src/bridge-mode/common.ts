export const DefaultBridgeServerPort = 3766;
export const DefaultLocalEndpoint = `http://127.0.0.1:${DefaultBridgeServerPort}`;
export const BridgeCallTimeout = 30000;

export enum BridgeEvent {
  Call = 'bridge-call',
  CallResponse = 'bridge-call-response',
  UpdateAgentStatus = 'bridge-update-agent-status',
  Message = 'bridge-message',
  Connected = 'bridge-connected',
  Refused = 'bridge-refused',
  ConnectNewTabWithUrl = 'connectNewTabWithUrl',
  ConnectCurrentTab = 'connectCurrentTab',
}

export interface BridgeConnectTabOptions {
  /**
   * If true, the page will always track the active tab.
   * @default true
   */
  forceSameTabNavigation?: boolean;
}

export enum MouseEvent {
  PREFIX = 'mouse.',
  Click = 'mouse.click',
  Wheel = 'mouse.wheel',
  Move = 'mouse.move',
  Drag = 'mouse.drag',
}

export enum KeyboardEvent {
  PREFIX = 'keyboard.',
  Type = 'keyboard.type',
  Press = 'keyboard.press',
}

export const BridgePageType = 'page-over-chrome-extension-bridge';

export const BridgeErrorCodeNoClientConnected = 'no-client-connected';

export interface BridgeCall {
  method: string;
  args: any[];
  response: any;
  callTime: number;
  responseTime: number;
  callback: (error: Error | undefined, response: any) => void;
  error?: Error;
}

export interface BridgeCallRequest {
  id: string;
  method: string;
  args: any[];
}

export interface BridgeCallResponse {
  id: string;
  response: any;
  error?: any;
}

export interface BridgeConnectedEventPayload {
  version: string;
}
