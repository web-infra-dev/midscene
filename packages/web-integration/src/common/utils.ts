import assert from 'node:assert';
import type { Buffer } from 'node:buffer';
import { randomUUID } from 'node:crypto';
import { readFileSync } from 'node:fs';
import path from 'node:path';
import type { ElementInfo } from '@/extractor';
import type { PlaywrightParserOpt, UIContext } from '@midscene/core';
import { getTmpFile } from '@midscene/core/utils';
import { findNearestPackageJson } from '@midscene/shared/fs';
import { base64Encoded, imageInfoOfBase64 } from '@midscene/shared/img';
import dayjs from 'dayjs';
import { WebElementInfo } from '../web-element';
import type { WebPage } from './page';

export type WebUIContext = UIContext<WebElementInfo> & {
  url: string;
};

export async function parseContextFromWebPage(
  page: WebPage,
  _opt?: PlaywrightParserOpt,
): Promise<WebUIContext> {
  assert(page, 'page is required');

  const url = page.url();
  const file = await page.screenshot();
  const screenshotBuffer = readFileSync(file);
  const screenshotBase64 = base64Encoded(file);
  const captureElementSnapshot = await page.getElementInfos();

  // align element
  const elementsInfo = await alignElements(
    screenshotBuffer,
    captureElementSnapshot,
    page,
  );

  const size = await imageInfoOfBase64(screenshotBase64);

  return {
    content: elementsInfo,
    size,
    screenshotBase64,
    url,
  };
}

export async function getExtraReturnLogic() {
  const pathDir = findNearestPackageJson(__dirname);
  assert(pathDir, `can't find pathDir, with ${__dirname}`);
  const scriptPath = path.join(pathDir, './dist/script/htmlElement.js');
  const elementInfosScriptContent = readFileSync(scriptPath, 'utf-8');
  return `${elementInfosScriptContent}midscene_element_inspector.webExtractTextWithPosition()`;
}

const sizeThreshold = 3;
async function alignElements(
  screenshotBuffer: Buffer,
  elements: ElementInfo[],
  page: WebPage,
): Promise<WebElementInfo[]> {
  const validElements = elements.filter((item) => {
    return (
      item.rect.height >= sizeThreshold && item.rect.width >= sizeThreshold
    );
  });
  const textsAligned: WebElementInfo[] = [];
  for (const item of validElements) {
    const { rect, id, content, attributes, locator } = item;
    textsAligned.push(
      new WebElementInfo({
        rect,
        locator,
        id,
        content,
        attributes,
        page,
      }),
    );
  }

  return textsAligned;
}

export function reportFileName(tag = 'web') {
  const dateTimeInFileName = dayjs().format('YYYY-MM-DD_HH-mm-ss-SSS');
  return `${tag}-${dateTimeInFileName}`;
}

export function printReportMsg(filepath: string) {
  console.log('Midscene - report file updated:', filepath);
}

/**
 * Get the current execution file name
 * @returns The name of the current execution file
 */
export function getCurrentExecutionFile(trace?: string): string | false {
  const error = new Error();
  const stackTrace = trace || error.stack;
  const pkgDir = process.cwd() || '';
  if (stackTrace) {
    const stackLines = stackTrace.split('\n');
    for (const line of stackLines) {
      if (
        line.includes('.spec.') ||
        line.includes('.test.') ||
        line.includes('.ts') ||
        line.includes('.js')
      ) {
        const match = line.match(/(?:at\s+)?(.*?\.(?:spec|test)\.[jt]s)/);
        if (match?.[1]) {
          const targetFileName = match[1]
            .replace(pkgDir, '')
            .trim()
            .replace('at ', '');
          return targetFileName;
        }
      }
    }
  }
  return false;
}

/**
 * Generates a unique cache ID based on the current execution file and a counter.
 *
 * This function creates a cache ID by combining the name of the current execution file
 * (typically a test or spec file) with an incrementing index. This ensures that each
 * cache ID is unique within the context of a specific test file across multiple executions.
 *
 * The function uses a Map to keep track of the index for each unique file, incrementing
 * it with each call for the same file.
 *
 * @returns {string} A unique cache ID in the format "filename-index"
 *
 * @example
 * // First call for "example.spec.ts"
 * generateCacheId(); // Returns "example.spec.ts-1"
 *
 * // Second call for "example.spec.ts"
 * generateCacheId(); // Returns "example.spec.ts-2"
 *
 * // First call for "another.test.ts"
 * generateCacheId(); // Returns "another.test.ts-1"
 */

const testFileIndex = new Map<string, number>();
export function generateCacheId(fileName?: string): string {
  let taskFile = fileName || getCurrentExecutionFile();
  if (!taskFile) {
    taskFile = randomUUID();
    console.warn(
      'Midscene - using random UUID for cache id. Cache may be invalid.',
    );
  }

  if (testFileIndex.has(taskFile)) {
    const currentIndex = testFileIndex.get(taskFile);
    if (currentIndex !== undefined) {
      testFileIndex.set(taskFile, currentIndex + 1);
    }
  } else {
    testFileIndex.set(taskFile, 1);
  }
  return `${taskFile}-${testFileIndex.get(taskFile)}`;
}
