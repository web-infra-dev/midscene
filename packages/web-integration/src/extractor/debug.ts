import { webExtractTextWithPosition } from '.';
import { setExtractTextWithPositionOnWindow } from './util';

console.log(webExtractTextWithPosition(document.body, true));
console.log(JSON.stringify(webExtractTextWithPosition(document.body, true)));
setExtractTextWithPositionOnWindow();
