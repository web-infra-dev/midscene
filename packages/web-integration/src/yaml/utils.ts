import assert from 'node:assert';
import yaml from 'js-yaml';

import type { MidsceneYamlScript } from '@midscene/core';
import type {
  MidsceneYamlFlowItem,
  MidsceneYamlFlowItemAIAction,
  MidsceneYamlFlowItemAIAssert,
  MidsceneYamlFlowItemAIQuery,
  MidsceneYamlFlowItemAIWaitFor,
  MidsceneYamlFlowItemSleep,
} from '@midscene/core';

function interpolateEnvVars(content: string): string {
  return content.replace(/\$\{([^}]+)\}/g, (_, envVar) => {
    const value = process.env[envVar.trim()];
    if (value === undefined) {
      throw new Error(`Environment variable "${envVar.trim()}" is not defined`);
    }
    return value;
  });
}

export function parseYamlScript(
  content: string,
  filePath?: string,
): MidsceneYamlScript {
  const interpolatedContent = interpolateEnvVars(content);
  const obj = yaml.load(interpolatedContent) as MidsceneYamlScript;
  const pathTip = filePath ? `, failed to load ${filePath}` : '';
  assert(obj.target, `property "target" is required in yaml script${pathTip}`);
  assert(
    typeof obj.target === 'object',
    `property "target" must be an object${pathTip}`,
  );
  assert(obj.tasks, `property "tasks" is required in yaml script${pathTip}`);
  assert(
    Array.isArray(obj.tasks),
    `property "tasks" must be an array${pathTip}`,
  );
  return obj;
}

export const flowItemBrief = (flowItem?: MidsceneYamlFlowItem) => {
  if (!flowItem) {
    return '';
  }

  const sliceText = (text?: string) => {
    const lengthLimit = 60;
    if (text && text.length > lengthLimit) {
      return `${text.slice(0, lengthLimit)}...`;
    }

    return text || '';
  };

  if (
    (flowItem as MidsceneYamlFlowItemAIAction).aiAction ||
    (flowItem as MidsceneYamlFlowItemAIAction).ai
  ) {
    const lastTip = (
      (flowItem as MidsceneYamlFlowItemAIAction).aiActionProgressTips || []
    ).at(-1);
    return `aiAction: ${sliceText(
      lastTip ||
        (flowItem as MidsceneYamlFlowItemAIAction).aiAction ||
        (flowItem as MidsceneYamlFlowItemAIAction).ai,
    )}`;
  }
  if ((flowItem as MidsceneYamlFlowItemAIAssert).aiAssert) {
    return `aiAssert: ${sliceText(
      (flowItem as MidsceneYamlFlowItemAIAssert).aiAssert,
    )}`;
  }
  if ((flowItem as MidsceneYamlFlowItemAIQuery).aiQuery) {
    return `aiQuery: ${sliceText((flowItem as MidsceneYamlFlowItemAIQuery).aiQuery)}`;
  }
  if ((flowItem as MidsceneYamlFlowItemAIWaitFor).aiWaitFor) {
    return `aiWaitFor: ${sliceText(
      (flowItem as MidsceneYamlFlowItemAIWaitFor).aiWaitFor,
    )}`;
  }
  if ((flowItem as MidsceneYamlFlowItemSleep).sleep) {
    return `sleep: ${(flowItem as MidsceneYamlFlowItemSleep).sleep}`;
  }
  return '';
};
