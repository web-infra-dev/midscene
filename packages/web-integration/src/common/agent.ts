import type { WebPage } from '@/common/page';
import type {
  AgentWaitForOpt,
  ExecutionDump,
  GroupedActionDump,
} from '@midscene/core';
import {
  groupedActionDumpFileExt,
  stringifyDumpData,
  writeLogFile,
} from '@midscene/core/utils';
import { PageTaskExecutor } from '../common/tasks';
import type { AiTaskCache } from './task-cache';
import { printReportMsg, reportFileName } from './utils';

export interface PageAgentOpt {
  testId?: string;
  cacheId?: string;
  groupName?: string;
  groupDescription?: string;
  cache?: AiTaskCache;
  /* if auto generate report, default true */
  generateReport?: boolean;
  /* if auto print report msg, default true */
  autoPrintReportMsg?: boolean;
}

export class PageAgent {
  page: WebPage;

  dump: GroupedActionDump;

  reportFile?: string;

  reportFileName?: string;

  taskExecutor: PageTaskExecutor;

  opts: PageAgentOpt;

  constructor(page: WebPage, opts?: PageAgentOpt) {
    this.page = page;
    this.opts = Object.assign(
      {
        generateReport: true,
        autoPrintReportMsg: true,
        groupName: 'Midscene Report',
        groupDescription: '',
      },
      opts || {},
    );
    this.dump = {
      groupName: this.opts.groupName!,
      groupDescription: this.opts.groupDescription,
      executions: [],
    };
    this.taskExecutor = new PageTaskExecutor(this.page, {
      cacheId: opts?.cacheId,
    });
    this.reportFileName = reportFileName(opts?.testId || 'web');
  }

  appendExecutionDump(execution: ExecutionDump) {
    const currentDump = this.dump;
    currentDump.executions.push(execution);
  }

  dumpDataString() {
    // update dump info
    this.dump.groupName = this.opts.groupName!;
    this.dump.groupDescription = this.opts.groupDescription;
    return stringifyDumpData(this.dump);
  }

  writeOutActionDumps() {
    const { generateReport, autoPrintReportMsg } = this.opts;
    this.reportFile = writeLogFile({
      fileName: this.reportFileName!,
      fileExt: groupedActionDumpFileExt,
      fileContent: this.dumpDataString(),
      type: 'dump',
      generateReport,
    });

    if (generateReport && autoPrintReportMsg) {
      printReportMsg(this.reportFile);
    }
  }

  async aiAction(taskPrompt: string) {
    const { executor } = await this.taskExecutor.action(taskPrompt);
    this.appendExecutionDump(executor.dump());
    this.writeOutActionDumps();

    if (executor.isInErrorState()) {
      const errorTask = executor.latestErrorTask();
      throw new Error(`${errorTask?.error}\n${errorTask?.errorStack}`);
    }
  }

  async aiQuery(demand: any) {
    const { output, executor } = await this.taskExecutor.query(demand);
    this.appendExecutionDump(executor.dump());
    this.writeOutActionDumps();

    if (executor.isInErrorState()) {
      const errorTask = executor.latestErrorTask();
      throw new Error(`${errorTask?.error}\n${errorTask?.errorStack}`);
    }
    return output;
  }

  async aiAssert(assertion: string, msg?: string) {
    const { output, executor } = await this.taskExecutor.assert(assertion);
    this.appendExecutionDump(executor.dump());
    this.writeOutActionDumps();

    if (!output?.pass) {
      const errMsg = msg || `Assertion failed: ${assertion}`;
      const reasonMsg = `Reason: ${output?.thought || '(no_reason)'}`;
      throw new Error(`${errMsg}\n${reasonMsg}`);
    }
  }

  async aiWaitFor(assertion: string, opt?: AgentWaitForOpt) {
    const { executor } = await this.taskExecutor.waitFor(assertion, {
      timeoutMs: opt?.timeoutMs || 15 * 1000,
      checkIntervalMs: opt?.checkIntervalMs || 3 * 1000,
      assertion,
    });
    this.appendExecutionDump(executor.dump());
    this.writeOutActionDumps();

    if (executor.isInErrorState()) {
      const errorTask = executor.latestErrorTask();
      throw new Error(`${errorTask?.error}\n${errorTask?.errorStack}`);
    }
  }

  async ai(taskPrompt: string, type = 'action') {
    if (type === 'action') {
      return this.aiAction(taskPrompt);
    }
    if (type === 'query') {
      return this.aiQuery(taskPrompt);
    }

    if (type === 'assert') {
      return this.aiAssert(taskPrompt);
    }

    throw new Error(
      `Unknown type: ${type}, only support 'action', 'query', 'assert'`,
    );
  }
}
