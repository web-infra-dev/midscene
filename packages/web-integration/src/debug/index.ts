import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import path from 'node:path';
import type { WebPage } from '@/common/page';
import type { ElementInfo } from '@/extractor';
import { NodeType } from '@/extractor/constants';
import {
  processImageElementInfo,
  resizeImg,
  saveBase64Image,
} from '@midscene/shared/img';

export async function generateExtractData(
  page: WebPage,
  targetDir: string,
  saveImgType?: {
    disableInputImage: boolean;
    disableOutputImage: boolean;
    disableOutputWithoutTextImg: boolean;
    disableResizeOutputImg: boolean;
    disableSnapshot: boolean;
  },
) {
  const file = await page.screenshot();
  const screenshotBuffer = readFileSync(file);

  const inputImgBase64 = screenshotBuffer.toString('base64');

  const {
    elementsPositionInfo,
    captureElementSnapshot,
    elementsPositionInfoWithoutText,
  } = await getElementInfos(page);

  const inputImagePath = path.join(targetDir, 'input.png');
  const outputImagePath = path.join(targetDir, 'output.png');
  const outputWithoutTextImgPath = path.join(
    targetDir,
    'output_without_text.png',
  );
  const resizeOutputImgPath = path.join(targetDir, 'resize-output.png');
  const snapshotJsonPath = path.join(targetDir, 'element-snapshot.json');

  const {
    compositeElementInfoImgBase64,
    compositeElementInfoImgWithoutTextBase64,
  } = await processImageElementInfo({
    elementsPositionInfo,
    elementsPositionInfoWithoutText,
    inputImgBase64,
  });

  const resizeImgBase64 = (await resizeImg(inputImgBase64)) as string;

  const existingSnapshot = existsSync(snapshotJsonPath)
    ? JSON.parse(readFileSync(snapshotJsonPath, 'utf-8'))
    : null;

  if (
    existingSnapshot &&
    JSON.stringify(existingSnapshot) === JSON.stringify(captureElementSnapshot)
  ) {
    console.log('skip save snapshot for ', targetDir);
    return;
  }

  if (!saveImgType?.disableSnapshot) {
    writeFileSyncWithDir(
      snapshotJsonPath,
      JSON.stringify(captureElementSnapshot, null, 2),
    );
  }
  if (!saveImgType?.disableInputImage) {
    await saveBase64Image({
      base64Data: inputImgBase64,
      outputPath: inputImagePath,
    });
  }
  if (!saveImgType?.disableOutputImage) {
    await saveBase64Image({
      base64Data: compositeElementInfoImgBase64,
      outputPath: outputImagePath,
    });
  }
  if (!saveImgType?.disableOutputWithoutTextImg) {
    await saveBase64Image({
      base64Data: compositeElementInfoImgWithoutTextBase64,
      outputPath: outputWithoutTextImgPath,
    });
  }
  if (!saveImgType?.disableResizeOutputImg) {
    await saveBase64Image({
      base64Data: resizeImgBase64,
      outputPath: resizeOutputImgPath,
    });
  }
}

export function generateTestDataPath(testDataName: string) {
  // `dist/lib/index.js` Is the default export path
  const modulePath = require
    .resolve('@midscene/core')
    .replace('dist/lib/index.js', '');
  const midsceneTestDataPath = path.join(
    modulePath,
    `tests/ai/evaluate/test-data/${testDataName}`,
  );

  return midsceneTestDataPath;
}

function ensureDirectoryExistence(filePath: string) {
  const dirname = path.dirname(filePath);
  if (existsSync(dirname)) {
    return;
  }
  ensureDirectoryExistence(dirname);
  mkdirSync(dirname);
}

type WriteFileSyncParams = Parameters<typeof writeFileSync>;

export function writeFileSyncWithDir(
  filePath: string,
  content: WriteFileSyncParams[1],
  options: WriteFileSyncParams[2] = {},
) {
  ensureDirectoryExistence(filePath);
  writeFileSync(filePath, content, options);
}

export async function getElementInfos(page: WebPage) {
  const captureElementSnapshot: Array<ElementInfo> =
    await page.getElementInfos();

  const elementsPositionInfo = captureElementSnapshot.map(
    (elementInfo, index) => {
      return {
        label: elementInfo.indexId?.toString() || index.toString(),
        x: elementInfo.rect.left,
        y: elementInfo.rect.top,
        width: elementInfo.rect.width,
        height: elementInfo.rect.height,
        attributes: elementInfo.attributes,
      };
    },
  );
  const elementsPositionInfoWithoutText = elementsPositionInfo.filter(
    (elementInfo) => {
      if (elementInfo.attributes.nodeType === NodeType.TEXT) {
        return false;
      }
      return true;
    },
  );
  return {
    elementsPositionInfo,
    captureElementSnapshot,
    elementsPositionInfoWithoutText,
  };
}
