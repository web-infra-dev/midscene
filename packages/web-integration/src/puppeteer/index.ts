import { PageAgent, type PageAgentOpt } from '@/common/agent';
import type { Page as PuppeteerPage } from 'puppeteer';
import { WebPage as PuppeteerWebPage } from './page';

export { WebPage as PuppeteerWebPage } from './page';
export class PuppeteerAgent extends PageAgent {
  constructor(page: PuppeteerPage, opts?: PageAgentOpt) {
    const webPage = new PuppeteerWebPage(page);
    super(webPage, opts);
  }
}
