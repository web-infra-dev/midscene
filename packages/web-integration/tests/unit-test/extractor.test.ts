import path, { join } from 'node:path';
import { parseContextFromWebPage } from '@/common/utils';
import { generateExtractData } from '@/debug';
import { describe, expect, it } from 'vitest';
import { launchPage } from '../ai/puppeteer/utils';

const pagePath = join(__dirname, './fixtures/extractor/index.html');
describe(
  'extractor',
  () => {
    it('basic', async () => {
      const { page, reset } = await launchPage(`file://${pagePath}`);

      const { content } = await parseContextFromWebPage(page);
      await generateExtractData(
        page,
        path.join(__dirname, 'fixtures/extractor'),
        {
          disableInputImage: false,
          disableOutputImage: false,
          disableOutputWithoutTextImg: true,
          disableResizeOutputImg: true,
          disableSnapshot: true,
        },
      );

      const list = content.map((item) => {
        return {
          content: item.content,
          attributes: item.attributes,
        };
      });
      expect(list).toMatchSnapshot();
      await reset();
    });
  },
  {
    timeout: 90 * 1000,
  },
);
