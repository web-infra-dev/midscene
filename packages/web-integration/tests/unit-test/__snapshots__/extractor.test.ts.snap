// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractor > basic 1`] = `
[
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Data Record",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "1970-01-01 19:25:01",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "User Name: Stella",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "ID",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Field 2",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Field 3",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Field 4",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Field 5",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "30S",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Kace Cervantes",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Aylin Sawyer",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Jefferson Kirby",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Skyla Jefferson",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "70U",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Florence Davenport",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Dariel Acevedo",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Ashlynn Delacruz",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Memphis Leal",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "3AY",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Crystal Newman",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Anderson Brown",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Charlotte Griffith",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Franklin Everett",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "YPG",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Kori Payne",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Edward Blevins",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Aila Gill",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Matthias Reed",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "ZEN",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Magnolia Duke",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Kalel Glover",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Alessia Barton",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Cassius Peck",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Form",
  },
  {
    "attributes": {
      "for": "name",
      "htmlTagName": "<label>",
      "nodeType": "FORM_ITEM Node",
    },
    "content": "Name:",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Name:",
  },
  {
    "attributes": {
      "htmlTagName": "<input>",
      "id": "J_input",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "Hello World This is Placeholder",
    },
    "content": "Hello World This is Placeholder",
  },
  {
    "attributes": {
      "nodeType": "BUTTON Node",
    },
    "content": "Click Me",
  },
  {
    "attributes": {
      "htmlTagName": "<label>",
      "nodeType": "FORM_ITEM Node",
    },
    "content": "Shape",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Shape",
  },
  {
    "attributes": {
      "htmlTagName": "<input>",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "You shouldn't see this placeholder.",
      "value": "Rectangle",
    },
    "content": "Rectangle",
  },
  {
    "attributes": {
      "htmlTagName": "<textarea>",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "this_is_a_textarea",
    },
    "content": "this_is_a_textarea",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "English",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "中文",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Tiếng Việt",
  },
  {
    "attributes": {
      "for": "options",
      "htmlTagName": "<label>",
      "nodeType": "FORM_ITEM Node",
    },
    "content": "Choose an option:",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "Choose an option:",
  },
  {
    "attributes": {
      "htmlTagName": "<select>",
      "id": "options",
      "name": "options",
      "nodeType": "FORM_ITEM Node",
    },
    "content": "Option 1",
  },
  {
    "attributes": {
      "class": ".gh-ar-hdn",
      "for": "gh-ac",
      "htmlTagName": "<label>",
      "nodeType": "FORM_ITEM Node",
    },
    "content": "输入搜索关键词",
  },
  {
    "attributes": {
      "nodeType": "TEXT Node",
    },
    "content": "输入搜索关键词",
  },
  {
    "attributes": {
      "aria-autocomplete": "list",
      "aria-expanded": "false",
      "aria-haspopup": "true",
      "aria-label": "搜索任何物品",
      "aria-owns": "ui-id-1",
      "autocapitalize": "off",
      "autocomplete": "off",
      "autocorrect": "off",
      "class": ".gh-tb.ui-autocomplete-input",
      "htmlTagName": "<input>",
      "id": "gh-ac",
      "maxlength": "300",
      "name": "_nkw",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "搜索任何物品",
      "role": "combobox",
      "size": "50",
      "spellcheck": "false",
      "type": "text",
    },
    "content": "搜索任何物品",
  },
  {
    "attributes": {
      "class": ".life-core-input-inner__wrapper.life-core-input-inner__wrapper-border.life-core-input-inner__wrapper-size-md.life-core-input-inner__wrapper-add-suffix",
      "htmlTagName": "<label>",
      "nodeType": "FORM_ITEM Node",
    },
    "content": "",
  },
  {
    "attributes": {
      "class": ".life-core-input.life-core-input-size-md",
      "htmlTagName": "<input>",
      "nodeType": "FORM_ITEM Node",
      "placeholder": "验证码",
      "tabindex": "0",
      "type": "text",
    },
    "content": "验证码",
  },
  {
    "attributes": {
      "class": ".life-core-checkbox.life-core-checkbox-md.src-pages-Login-components-LoginCard-index-module__checkbox--Npo2V--212e2",
      "htmlTagName": "<label>",
      "nodeType": "FORM_ITEM Node",
    },
    "content": "",
  },
]
`;
